API/utils/Model.php

```diff
@@ -811,6 +811,121 @@ function RemoveAdmin($email) {
+     * @param string $audioRecordingFilename
+     * @param string $audioData
+     * @return int
+     */
+    public function UpdateWordLinkRecording($recordingId, $projectId, $androidId, $term, $textBackTranslation, $audioRecordingFilename, $audioData)
+    {
+         try {
+             # check a recording exists for given recordingId, term & projectId
+             $sql = "SELECT id, fileName FROM WordLinkRecordings WHERE id = ? AND term = ? AND projectId = ?;";
+             $stmt = $this->PrepareAndExecute($sql, array($recordingId, $term, $projectId));
+
+             $recording = $this->FetchArray($stmt);
+             $this->FreeStmt($stmt);
+
+             if (!is_array($recording)) {
+                 RespondWithError(400, "No WordLinkRecordings found for given recordingID: '$recordingId', term: '$term' and androidId: '$androidId'");
+             }
+
+             $existingRecodingFileName = $recording['fileName'];
+
+             # generate random file for saving uploaded recording file content
+             $generatedFileName = $this->GenerateRecordingFileName($audioRecordingFilename);
+
+             # update WordLinkRecordings
+             $this->conn->beginTransaction();
+             $updateSql =  "UPDATE WordLinkRecordings
+                            SET textBackTranslation = ?, fileName = ? WHERE id = ?";
+
+             $stmt = $this->PrepareAndExecute($updateSql, array($textBackTranslation, $generatedFileName, $recordingId));
+             $this->FreeStmt($stmt);
+
+             $this->conn->commit();
+             $directory = "Projects/$androidId/WordLinks";
+             PutFile($directory, $generatedFileName, $audioData);
+
+             # delete existing recording file
+             $existingRecodingFilepath = "{$GLOBALS['filesRoot']}/$directory/$existingRecodingFileName";
+
+             error_log(sprintf("Deleting old recording file '%s'", $existingRecodingFilepath));
+             if(!unlink($existingRecodingFilepath)) {
+                 throw new \Exception("Failed to deleted existing recording file: $existingRecodingFilepath");
+             }
+
+             return $recordingId;
+         } catch (\Exception $e) {
+             error_log($e->getMessage());
+             $this->conn->rollBack();
+             RespondWithError(500, $e->getMessage());
+         }
+    }
+
+    /**
+     * @throws DBException
+     */
+    function GetProjectIdId($androidId) {
+        $sql = "SELECT id FROM Projects WHERE androidId = ?";
+        $stmt = $this->PrepareAndExecute($sql, array($androidId));
+
+        if ($project = $this->FetchArray($stmt)) {
+            $this->FreeStmt($stmt);
+            return $project['id'];
+        }
+        return null;
+    }
+
+    function GenerateRecordingFileName($audioRecordingFilename): string
+    {
+        # extract file extension and verify if it's a valid one
+        $recordingFileExtension = pathinfo($audioRecordingFilename, PATHINFO_EXTENSION);
```

Comment:
Can you validate that extension is ascii alphanumeric only? e.g. regex `[A-Za-z0-9]`. Doesn't seem like pathinfo already does that.

Reply:
Let's do length of 1 to 100. I don't think any reasonable extension is longer than 100.